### PR TITLE
Internal PartitionKey: Adds ability to parse out partition key from document.

### PR DESCRIPTION
**Pull request by bchong95**
# Internal PartitionKey: Adds ability to parse out partition key from document.

## Description

Compute needs the ability to parse out the partition key from a document. Right now this logic is very internal to the SDK, so we are refactoring it out here so users and the SDK can call into it.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosContainerTests.cs [hunk 1]
```diff
@@ -1459,9 +1459,6 @@ private void ValidateCreateContainerResponseContract(ContainerResponse container
             Assert.IsTrue(containerSettings.LastModified.HasValue);
 
             Assert.IsNotNull(containerSettings.PartitionKeyPath);
-            Assert.IsNotNull(containerSettings.PartitionKeyPathTokens);
-            Assert.AreEqual(1, containerSettings.PartitionKeyPathTokens[0].Count);
-            Assert.AreEqual("id", containerSettings.PartitionKeyPathTokens[0][0]);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosContainerTests.cs:1464]: Can we keep this validation using the PartitionKeyPath property?

File: Microsoft.Azure.Cosmos/src/Resource/Settings/ContainerProperties.cs [hunk 2]
```diff
@@ -622,49 +618,6 @@ internal ContainerProperties(string id, PartitionKeyDefinition partitionKeyDefin
 
         internal bool HasPartitionKey => this.PartitionKey != null;
 
-        internal IReadOnlyList<IReadOnlyList<string>> PartitionKeyPathTokens
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/ContainerProperties.cs:625]: Is this not being used by anyone internal?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerInternal.cs [hunk 3]
```diff
@@ -46,9 +46,6 @@ internal abstract class ContainerInternal : Container
             bool forceRefresh = false,
             CancellationToken cancellationToken = default);
 
-        public abstract Task<IReadOnlyList<IReadOnlyList<string>>> GetPartitionKeyPathTokensAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerInternal.cs:49]: @SrinikhilReddy this seems to conflict with PR https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1952. Can you please reivew?

File: Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs [hunk 4]
```diff
@@ -81,6 +83,35 @@ public PartitionKeyBuilder AddNoneType()
...
+                throw new ArgumentNullException(nameof(cosmosElement));
+            }
+
+            return cosmosElement switch
+            {
+                CosmosString cosmosString => this.Add(cosmosString.Value),
+                CosmosNumber cosmosNumber => this.Add(Number64.ToDouble(cosmosNumber.Value)),
+                CosmosBoolean cosmosBoolean => this.Add(cosmosBoolean.Value),
+                CosmosNull _ => this.AddNullValue(),
+                _ => throw new ArgumentException(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs:107]: What about undefined?

File: Microsoft.Azure.Cosmos/src/Resource/Settings/ContainerProperties.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/ContainerProperties.cs:626]: No it's was only there for the partition key parsing step.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosContainerTests.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosContainerTests.cs:1465]: There is no Tokens so nothing to validate.

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerInternal.cs (hunk 3 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerInternal.cs:58]: The PR has been open for sometime.

File: Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs:107]: Undefined is null and is handled by the caller.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 5]
```diff
@@ -214,6 +220,66 @@ internal static bool TryParseJsonString(string partitionKeyString, out Partition
...
+                // In the future use spans to avoid an allocation here
+                // For make the path a tokenized type.
+                string[] tokens = path.Split(PartitionKey.partitionKeyTokenDelimeter, StringSplitOptions.RemoveEmptyEntries);
+
+                static bool TryGetPartitionKeyValueFromTokens(
+                    CosmosObject pathTraversalSoFar,
+                    ReadOnlySpan<string> tokens,
+                    out CosmosElement partitionKeyValue)
+                {
+                    // Note: this code is technically wrong, since a user could have a number index in the path.
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs:256]: Can a bug be filed for this?

File: Microsoft.Azure.Cosmos/src/Resource/Settings/ContainerProperties.cs [hunk 6]
```diff
@@ -623,49 +610,6 @@ internal ContainerProperties(string id, PartitionKeyDefinition partitionKeyDefin
...
-                {
-                    throw new NotImplementedException("PartitionKey extraction with composite partition keys not supported.");
-                }
-
-                if (this.PartitionKey.Kind != Documents.PartitionKind.MultiHash && this.PartitionKeyPath == null)
-                {
-                    throw new ArgumentOutOfRangeException($"Container {this.Id} is not partitioned");
-                }
-
-#if PREVIEW
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/ContainerProperties.cs:650]: I don't see this on the new code?

**Pull request closed without merging**